src/nexusrpc/handler/_core.py

```diff
@@ -0,0 +1,463 @@
+
+    Instances of this class are created automatically from user service handler instances
+    on creation of a Handler instance, at Nexus handler start time. While the user's class
+    defines operation handlers as factory methods to be called at handler start time, this
+    class contains the :py:class:`OperationHandler` instances themselves.
+
+    You may create instances of this class manually and pass them to the Handler
+    constructor, for example when programatically creating Nexus service implementations.
+    """
+
+    service: nexusrpc.ServiceDefinition
+    operation_handlers: dict[str, OperationHandler[Any, Any]]
+
+    @classmethod
+    def from_user_instance(cls, user_instance: Any) -> Self:
+        """Create a :py:class:`ServiceHandler` from a user service instance."""
+
+        service = getattr(user_instance.__class__, "__nexus_service__", None)
+        if not isinstance(service, nexusrpc.ServiceDefinition):
+            raise RuntimeError(
+                f"Service '{user_instance}' does not have a service definition. "
+                f"Use the :py:func:`@nexusrpc.handler.service_handler` decorator on your class to define "
+                f"a Nexus service implementation."
+            )
+        op_handlers = {
+            name: factory(user_instance)
+            for name, factory in collect_operation_handler_factories(
+                user_instance.__class__, service
+            ).items()
+        }
+        return cls(
+            service=service,
+            operation_handlers=op_handlers,
+        )
+
+    def _get_operation_handler(self, operation: str) -> OperationHandler[Any, Any]:
+        """Return an operation handler, given the operation name."""
+        if operation not in self.service.operations:
+            msg = (
+                f"Nexus service definition '{self.service.name}' has no operation '{operation}'. "
+                f"There are {len(self.service.operations)} operations in the definition."
+            )
+            if self.service.operations:
+                msg += f": {', '.join(sorted(self.service.operations.keys()))}"
+            msg += "."
+            raise HandlerError(msg, type=HandlerErrorType.NOT_FOUND)
+        operation_handler = self.operation_handlers.get(operation)
+        if operation_handler is None:
+            # This should not be possible. If a service definition was supplied then
+            # this was checked; if not then the definition was generated from the
+            # operation handlers.
+            msg = (
+                f"Nexus service implementation '{self.service.name}' has no handler for operation '{operation}'. "
+                f"There are {len(self.operation_handlers)} available operation handlers"
+            )
+            if self.operation_handlers:
+                msg += f": {', '.join(sorted(self.operation_handlers.keys()))}"
+            msg += "."
```

Comment:
Is it common to put periods at the end of error messages in Python? Doesn't seem like the standard library does this.

Reply:
There are two sentences here so it feels necessary to me.